File: src/ros2_bridge/mappers/nav2_msgs/behavior_tree_log.rs
//! Typed mapper for `nav2_msgs/msg/BehaviorTreeLog`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn behavior_tree_log_to_bus(msg: ros_env::nav2_msgs::msg::BehaviorTreeLog) -> crate::nav2_msgs::msg::v1::BehaviorTreeLog {
    crate::nav2_msgs::msg::v1::BehaviorTreeLog {
        timestamp: Some(crate::ros2_bridge::mappers::builtin_interfaces::time::time_to_bus(msg.timestamp)),
        event_log: msg.event_log.into_iter().map(crate::ros2_bridge::mappers::nav2_msgs::behavior_tree_status_change::behavior_tree_status_change_to_bus).collect(),
    }
}

pub(crate) fn behavior_tree_log_to_ros(bus: crate::nav2_msgs::msg::v1::BehaviorTreeLog) -> ros_env::nav2_msgs::msg::BehaviorTreeLog {
    ros_env::nav2_msgs::msg::BehaviorTreeLog {
        timestamp: crate::ros2_bridge::mappers::builtin_interfaces::time::time_to_ros(bus.timestamp.unwrap_or_default()),
        event_log: bus.event_log.into_iter().map(crate::ros2_bridge::mappers::nav2_msgs::behavior_tree_status_change::behavior_tree_status_change_to_ros).collect(),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct Nav2MsgsBehaviorTreeLogMapper;

impl TypedTopicMapper for Nav2MsgsBehaviorTreeLogMapper {
    type Ros = ros_env::nav2_msgs::msg::BehaviorTreeLog;
    type Bus = crate::nav2_msgs::msg::v1::BehaviorTreeLog;

    fn type_name(&self) -> &'static str {
        "nav2_msgs/msg/BehaviorTreeLog"
    }

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(behavior_tree_log_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(behavior_tree_log_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/nav2_msgs/behavior_tree_log.py
"""Generated mapper for `nav2_msgs/msg/BehaviorTreeLog`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.builtin_interfaces.time import time_to_bus, time_to_ros
from robot_bus.ros2_bridge.mappers.nav2_msgs.behavior_tree_status_change import behavior_tree_status_change_to_bus, behavior_tree_status_change_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.nav2_msgs.msg.v1 import BehaviorTreeLog as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def behavior_tree_log_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.timestamp.CopyFrom(time_to_bus(msg.timestamp))
    bus.event_log.extend([behavior_tree_status_change_to_bus(x) for x in msg.event_log])
    return bus


def behavior_tree_log_to_ros(bus):
    from nav2_msgs.msg import BehaviorTreeLog as RosMsg

    out = RosMsg()
    out.timestamp = time_to_ros(bus.timestamp)
    out.event_log = [behavior_tree_status_change_to_ros(x) for x in bus.event_log]
    return out


class Nav2MsgsBehaviorTreeLogMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from nav2_msgs.msg import BehaviorTreeLog as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return behavior_tree_log_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return behavior_tree_log_to_ros(bus)
